### PR TITLE
Add EntityNamePicker component to validate entity names in templates

### LINKED ISSUE #6687: scaffolder: sample templates should validate component name properly
## Expected Behavior

Components created through our sample templates should always be possible to register in the catalog using the default validation rules.

## Current Behavior

The validation of the entity name in our sample templates is too lax, for example allowing spaces. This will break the catalog registration step when creating the new component.

## Possible Solution

Validate the name field in our sample templates using the same Regex as entity names are validated by, for example here: https://github.com/backstage/backstage/blob/13f5d78b0bb079030c9c8d1d58cea2bb6bacfa38/plugins/scaffolder-backend/sample-templates/react-ssr-template/template.yaml#L20

## Steps to Reproduce

1. Try to create a component called `A B C`
2. Wait for scaffolder job to complete, which it will do succesfully
3. Try to find the component in the catalog, it's not there. Check backend logs for error: `Policy check failed; caused by Error: "metadata.name" is not valid; expected a string that is sequences of [a-zA-Z0-9] separated by any of [-_.], at most 63 characters in total but found "A B C"`

### PR DESCRIPTION
**Pull request by mtlewis**
## Hey, I just made a Pull Request!

Closes #6687.

- Adds an `EntityNamePicker` component that reuses the entity name validation from the catalog
- Adds this field to the default scaffolder field extensions

### To do 

- [x] Open PR to update [software-templates](https://github.com/backstage/software-templates) examples to use EntityNamePicker - https://github.com/backstage/software-templates/pull/5
- [x] Changeset

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: .changeset/happy-lies-wink.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+- Adds a new generic field `TextValuePicker` to be used when writing custom field extensions that use a standard UI with custom validation. An example of doing this can be found in `packages/app/src/components/scaffolder/customScaffolderExtensions.tsx`.
+
+## `@backstage/scaffolder-backend
+
+- Updates the sample templates to use the new `EntityNamePicker` field, and cleans up some unused field configuration.
+
+## `example-app`
+
+- Adds an example of providing a custom field extension.
```
Comment (OrkoHunter) [inline .changeset/happy-lies-wink.md:34]: We could split this into separate changeset files, looking at the divisions this one has using `##`. Or else, this whole thing will get copied over in all the `CHANGELOG` files. Also, `example-app` one is not needed, since we don't release the package!

File: plugins/scaffolder-backend/sample-templates/create-react-app/template.yaml [hunk 2]
```diff
@@ -20,6 +20,7 @@ spec:
           title: Name
           type: string
           description: Unique name of the component
+          ui:field: EntityNamePicker
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/sample-templates/create-react-app/template.yaml:23]: Can't update the templates before the UI component has been released.

Comment (mtlewis) [reply plugins/scaffolder-backend/sample-templates/create-react-app/template.yaml:23]: Makes sense! Will be nice to get rid of these templates soon. Will split this off :)

File: .changeset/happy-lies-wink.md (hunk 1 above)
Comment (mtlewis) [reply .changeset/happy-lies-wink.md:34]: Makes sense! I'll add a second changeset. Perhaps it would be useful for `example-app` to be omitted from the list of packages in the `yarn changeset` prompt 🤔.

File: plugins/scaffolder-backend/sample-templates/create-react-app/template.yaml (hunk 2 above)
Comment (mtlewis) [reply plugins/scaffolder-backend/sample-templates/create-react-app/template.yaml]: #6726.

File: .changeset/happy-lies-wink.md (hunk 1 above)
Comment (mtlewis) [reply .changeset/happy-lies-wink.md]: Changeset updated - the other half ended up in #6726.

Comment (mtlewis) [reply .changeset/happy-lies-wink.md]: Also - opened https://github.com/backstage/backstage/issues/6727 to look into omitting packages from the prompt in `yarn changeset`.

File: packages/app/src/components/scaffolder/customScaffolderExtensions.tsx [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type { FieldValidation } from '@rjsf/core';
+import {
+  createScaffolderFieldExtension,
+  TextValuePicker,
+  scaffolderPlugin,
```
Comment (benjdlambert) [inline packages/app/src/components/scaffolder/customScaffolderExtensions.tsx:21]: This is an interesting pattern. So you don't need to create your own plugin context to do the provide. Think I will put this in my docs for creating your own extensions

**Review by benjdlambert — APPROVED** [review]

File: plugins/scaffolder/src/components/fields/EntityNamePicker/validation.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { FieldValidation } from '@rjsf/core';
+import { KubernetesValidatorFunctions } from '@backstage/catalog-model';
```
Comment (freben) [inline plugins/scaffolder/src/components/fields/EntityNamePicker/validation.ts:18]: I just want to mention that this is only one possible rule set for names. Some orgs change this, and will want it to act accordingly in the frontend as well.

**Review by freben — APPROVED** [review]

File: plugins/scaffolder/src/components/fields/EntityNamePicker/validation.ts (hunk 4 above)
Comment (mtlewis) [reply plugins/scaffolder/src/components/fields/EntityNamePicker/validation.ts:18]: Totally! I believe @benjdlambert has some WIP documentation for scaffolder field extensions. Once that lands, we can extend it to document the steps required to configure an EntityNamePicker with different validation rules.
